runners/flink/src/main/java/org/apache/beam/runners/flink/FlinkStreamingPortablePipelineTranslator.java

```diff
@@ -0,0 +1,443 @@
+    for (PipelineNode.PTransformNode transform : p.getTopologicallyOrderedTransforms()) {
+      urnToTransformTranslator.getOrDefault(
+              transform.getTransform().getSpec().getUrn(), this::urnNotFound)
+              .translate(transform.getId(), pipeline, context);
+    }
+
+  }
+
+  private void urnNotFound(
+          String id, RunnerApi.Pipeline pipeline,
+          FlinkStreamingPortablePipelineTranslator.TranslationContext context) {
+    throw new IllegalArgumentException(
+            String.format("Unknown type of URN %s for PTrasnform with id %s.",
+                    pipeline.getComponents().getTransformsOrThrow(id).getSpec().getUrn(),
+                    id));
+  }
+
+  private <K, V> void translateReshuffle(
+      String id,
+      RunnerApi.Pipeline pipeline,
+      StreamingTranslationContext context) {
+    RunnerApi.PTransform transform = pipeline.getComponents().getTransformsOrThrow(id);
+    DataStream<WindowedValue<KV<K, V>>> inputDataSet =
+        context.getDataStreamOrThrow(
+            Iterables.getOnlyElement(transform.getInputsMap().values()));
+    context.addDataStream(Iterables.getOnlyElement(transform.getOutputsMap().values()),
+        inputDataSet.rebalance());
+  }
+
+  private <T>  void translateFlatten(
+          String id,
+          RunnerApi.Pipeline pipeline,
+          StreamingTranslationContext context) {
+    Map<String, String> allInputs =
+            pipeline.getComponents().getTransformsOrThrow(id).getInputsMap();
+
+    if (allInputs.isEmpty()) {
+
+      // create an empty dummy source to satisfy downstream operations
+      // we cannot create an empty source in Flink, therefore we have to
+      // add the flatMap that simply never forwards the single element
+      DataStreamSource<String> dummySource =
+              context.getExecutionEnvironment().fromElements("dummy");
+
+      DataStream<WindowedValue<T>> result =
+              dummySource
+                      .<WindowedValue<T>>flatMap(
+                              (s, collector) -> {
+                                // never return anything
+                              })
+                      .returns(
+                              new CoderTypeInformation<>(
+                                      WindowedValue.getFullCoder(
+                                              (Coder<T>) VoidCoder.of(),
+                                              GlobalWindow.Coder.INSTANCE)));
+      context.addDataStream(Iterables.getOnlyElement(
+              pipeline.getComponents().getTransformsOrThrow(id).getOutputsMap().values()), result);
+    } else {
+      DataStream<T> result = null;
+
+      // Determine DataStreams that we use as input several times. For those, we need to uniquify
+      // input streams because Flink seems to swallow watermarks when we have a union of one and
+      // the same stream.
+      Map<DataStream<T>, Integer> duplicates = new HashMap<>();
```

Comment:
Consider using a Multiset (https://google.github.io/guava/releases/19.0/api/docs/com/google/common/collect/Multiset.html) as it provides the frequency count feature already.